Editor.next and Editor.previous now return the first match

tomdng opened the pull request:
#### Is this adding or improving a _feature_ or fixing a _bug_?

Fixing a bug

#### What's the new behavior?

`Editor.next` and `Editor.previous` now return the **1st match** instead of the 2nd match.

#### How does this change work?

This changes works by using `Editor.before` and `Editor.after` to find the point before/after the current position passed into `Editor.previous` and `Editor.next` respectively.

However, this implementation relies in a modification to `Editor.before`, `Editor.after`, `Editor.positions`, and `Editor.string` where all four functions now accept an optional voids parameter so that you can iterate void nodes instead of treating them as a single point. By default the option is set to false to match original behavior, but passing in `voids: true` to these functions should enable the new behavior that make this implementation work.

Finally I had to adjust two test cases because of a new issue I found with `Editor.before`. It throws an error it finds a location without a text node that exists before the current location. So using `Editor.before` to resolve `Editor.previous` caused two tests cases to throw an error.

This only occurs during two `moveNodes` test cases because in the process of moving a node out of the first location of the editor, the Editor hasn't fully normalized that location yet, so when `Editor.before` is called, it then calls `Editor.start` (which ultimately calls `Editor.point` with the `edge: 'start'` option) which then throws an error because there is no start text node.

I know that changing the test cases is not recommended, as it introduces different behavior, so would anyone have a better method to get `Editor.next` and `Editor.previous` to return the 1st match while passing the original test case? Otherwise we could expose another function like `Editor.before` and `Editor.after` that doesn't end up throwing an error when running on un-normalized nodes. I would appreciate any help with this!

#### Have you checked that...?

- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)

#### Does this fix any issues or need any specific reviewers?

Fixes: #3791 
Fixes: #3590

BrentFarese commented on packages/slate/src/interfaces/editor.ts:
```diff
@@ -121,16 +121,21 @@ export const Editor = {
+    for (const p of Editor.positions(editor, {
+      ...options,
+      at: range,
+      voids,
```
Since we are spreading `...options` on 135 we can get rid of 137.

```suggestion
```

BrentFarese commented on packages/slate/src/interfaces/editor.ts:
```diff
@@ -1107,8 +1126,18 @@ export const Editor = {
+
+    // The search location is from the start of the document to the path of
+    // the point before the location passed in
     const span: Span = [from, to]
```
Will have to change this line if you accept my suggestion above.

```suggestion
    const span: Span = [pointBeforeLocation.path, to]
```

BrentFarese reviewed (changes requested):
Looking good @tomdng.  I think 1 thing missing from this PR is that we should test the new behavior of `Editor.positions` with void nodes.  Can you write a test or 2 for that use case?  Similarly, we can consider a test case or 2 for `voids = true` for `Editor.before` and `Editor.after`, if we want test coverage to be solid.

Thanks!

tomdng commented:
Thank you @BrentFarese! I went ahead and added the test cases to `Editor.before`, `Editor.after`, and `Editor.position` with the `voids: true` option. I also went ahead and added an additional test case to `Editor.string` because that also has the new `voids: true` behavior.

Let me know if you want more/less test cases, especially for `Editor.positions` because it looks like there's a ton of test cases for the non void elements.

BrentFarese reviewed (approved):
Test coverage looks great!  Thanks @tomdng.

BobElDevil commented:
@BrentFarese  FYI, noticed this while putting up my own PR. Due to the changes done here https://github.com/ianstormtaylor/slate/issues/3770 PR's should now be targeted towards 'main' not 'master', so this PR got merged into the 'wrong' branch. Probably want to point all open PRs to the new branch and re-merge these changes

lukesmurray commented:
@BrentFarese string is missing the new parameter options in the editor interface which leads to typescript issues.

https://github.com/ianstormtaylor/slate/blob/db3f52517fb1e1dd81301fb084cfaddc5d73f62f/packages/slate/src/interfaces/editor.ts#L249

BobElDevil commented:
I actually opened up a PR for this recently https://github.com/ianstormtaylor/slate/pull/4009

lukesmurray commented:
@BobElDevil amazing thank you!

BrentFarese commented:
> I actually opened up a PR for this recently #4009

Wish I had seen this!  #4014 is a duplicate of this.  1 of the PRs will get merged and this will be fixed.  Thanks!
